### PR TITLE
script to subscribe event proof and submit

### PR DESCRIPTION
**Pull request by KarishmaBothara**
Have tested this, by adding validator and calling forceNewEra, waited for next session

```
yarn subscribeEventProof

2021-09-10 12:15:01        API/INIT: RPC methods not decorated: genericAsset_getBalance, nft_getCollectionInfo, nft_getTokenInfo, system_unstable_networkState
Connecting to CENNZnet bridge contract...
CENNZnet bridge deployed to: 0x5FbDB2315678afecb367f032d93F642f64180aa3

data:: {
  digest: '0x4f12d20d0370f1416186490e2c28723a7d33b0f64d2e621f528f5fa3c2c32108',
  eventId: '0',
  validatorSetId: '0',
  signatures: [
    '0x5458848bb0524b5ca5efc21707aabfe0143c96498e1d02127a542465bccfa82b271a803005f3615028ef1a2e3ac2f0f435251fbc2628acbc088df99b8bac3f9301'
  ]
}
0x0404dad6fc9c291c68498de501c6d6d17bfe28aee69cfbf71b2cc849caafcb0159dadfbd20856d9ed9d2f77742372d4fcac775b78878d2be96455162b19623f780
{
  hash: '0x2b531b1d811d3541b887a29e59d925d7348d1d0d59bad21cf999b60740b962aa',
  type: 0,
  accessList: null,
  blockHash: '0xbdb8376fb5a16aa58f28ad1b10534006ece1d2b39cad028b65e705f738fed6ee',
  blockNumber: 12,
  transactionIndex: 0,
  confirmations: 1,
  from: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266',
  gasPrice: BigNumber { _hex: '0x01dcd65000', _isBigNumber: true },
  gasLimit: BigNumber { _hex: '0x019e1e', _isBigNumber: true },
  to: '0x5FbDB2315678afecb367f032d93F642f64180aa3',
  value: BigNumber { _hex: '0x00', _isBigNumber: true },
  nonce: 11,
  data: '0xb8f480c600000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000000100000000000000000000000058dad74c38e9c4738bf3471f6aac6124f862faf50000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000001200000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000001c00000000000000000000000000000000000000000000000000000000000000015458848bb0524b5ca5efc21707aabfe0143c96498e1d02127a542465bccfa82b0000000000000000000000000000000000000000000000000000000000000001271a803005f3615028ef1a2e3ac2f0f435251fbc2628acbc088df99b8bac3f93',
  r: '0xb33ac473befa6b17a7bdbf243af8ff33d0e97b2ca492b2e4bb29b370fb6d823c',
  s: '0x4e8202d37d6025d7a5a039dd7dbccfb9060582c13cd951e314abf946d58cfdca',
  v: 62709,
  creates: null,
  chainId: 31337,
  wait: [Function]
}

```
The script works, and updates the validators set and signature when AuthoritySetChange event is emitted.

### CODE REVIEW

File: scripts/subscribeEventProof.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                    const { eventId, validatorSetId, signatures } = eventProof;
+                    const signature = ethers.utils.hexlify(signatures[0]);
+                    const sig = ethers.utils.splitSignature(signature);
+                    const cennznetEventProof =
+                            { eventId,
+                              validatorSetId,
+                              v: [sig.v],
+                              r: [sig.r],
+                              s: [sig.s]
+                            };
```
Comment (holygits) [inline scripts/subscribeEventProof.js:38]: should expand this part to work for `n` signatures

Comment (KarishmaBothara) [reply scripts/subscribeEventProof.js]: done

File: scripts/subscribeEventProof.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        .subscribeFinalizedHeads(async (head) => {
+           const blockNumber = head.number.toNumber();
+           console.log('At blocknumber:',blockNumber);
+           const blockHash = head.hash.toString();
+           const events = await api.query.system.events.at(blockHash);
+           events.map(async ({event}) => {
+                const { section, method, data } = event;
+                if (section === 'ethBridge' && method === 'AuthoritySetChange') {
+                    const eventIdFound = data.toHuman()[0];
+                    const eventProof = (await api.rpc.ethy.getEventProof(eventIdFound)).toJSON().EventProof;
```
Comment (holygits) [inline scripts/subscribeEventProof.js:59]: what happens if this comes back empty? e.g proof isn't read yet

Comment (KarishmaBothara) [reply scripts/subscribeEventProof.js:59]: Once we deploy on rata.. may be I can start the script to see what happens.. I haven't really encountered a scenario where the proof isn't ready..

Comment (holygits) [reply scripts/subscribeEventProof.js:59]: it will definitely happen on real network as the proof takes sometime to be signed by all the active validators

Comment (KarishmaBothara) [reply scripts/subscribeEventProof.js]: have updated the script.. but I am keen to test it on rata to see if it works

File: scripts/e2e.js [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    const transaction1 = api.tx.erc20Peg.activateDeposits(true);
+    const transaction2 = api.tx.erc20Peg.activateWithdrawals(true);
+    const transaction3 = api.tx.erc20Peg.setContractAddress(peg.address);
+    const transaction4 = api.tx.ethBridge.setEventConfirmations(0); // Hardhat only makes blocks when txs are sent
+    const batchBridgeActivationEx = api.tx.utility.batch([
+        transaction1,
+        transaction2,
+        transaction3,
+        transaction4
+    ]);
```
Comment (holygits) [inline scripts/e2e.js:54]: could you add a comment about this only being necessary for a local chain, where it's not configured

File: scripts/e2e.js [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        tokenAddress: token.address,
+        amount: depositAmount,
+        beneficiary: cennznetAddress
+    };
+
+    console.log('testTokenId::', testTokenId.toString());
+
+    /***************** Make Deposit on CENNZnet **********************/
+
+    let nonce = await api.rpc.system.accountNextIndex(alice.address);
```
Comment (holygits) [inline scripts/e2e.js:78]: comment about `eventConfirmations = 0` so no need to wait

File: scripts/e2e.js [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    await new Promise(async (resolve, reject) => {
+        await api.tx.erc20Peg.depositClaim(depositTxHash, claim).signAndSend(alice, {nonce}, async ({status, events}) => {
+            if (status.isInBlock) {
+                for (const {event: {method, section, data}} of events) {
+                    console.log('\t', `: ${section}.${method}`, data.toString());
+                    if (section === 'erc20Peg' && method == 'Erc20Claim') {
+                        console.log('*******************************************');
```
Comment (holygits) [inline scripts/e2e.js:85]: this only means the claim started.
need to check future block events for `Erc20ClaimSuccess` with the matching Id
also best to match on the `sender` from the event is the same also

File: scripts/e2e.js (hunk 4 above)
Comment (holygits) [reply scripts/e2e.js:84]: bump

File: scripts/e2e.js [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+                eventProof = versionedEventProof.EventProof;
+                console.log('Event proof found;::', eventProof);
+                unsubHeads();
+                resolve();
+            }
+        });
+    });
+
+    /***************** Make Withdrawal on ETHEREUM **********************/
+    // Activate withdrawals
```
Comment (holygits) [inline scripts/e2e.js:159]: can we do this in contract setup step

File: package.json [hunk 7]
```diff
@@ -6,12 +6,16 @@
...
+    "erc20_e2e": "hardhat run --network localhost scripts/e2e.js",
     "eth_e2e": "hardhat run --network localhost scripts/ethEndToEndTest.js",
     "deposit": "hardhat run --network localhost scripts/deposit.js",
     "withdraw": "hardhat run --network localhost scripts/withdraw.js",
     "setAssetMeta": "node scripts/setAssetMeta.js",
     "setAssetMetaViaJSON": "node scripts/setMetadataViaStorage.js",
     "startEthNode": "./scripts/startEthNode.sh",
+    "validatorRelayer_local": "hardhat run --network localhost scripts/subscribeEventProof.js",
+    "validatorRelayer_ropsten": "hardhat run --network ropsten scripts/subscribeEventProof.js",
+    "validatorRelayer_mainnet": "hardhat run --network mainnet scripts/subscribeEventProof.js",
```
Comment (holygits) [inline package.json:18]: hmm I think can do:
```suggestion
    "validatorRelayer": "hardhat run --network $1 scripts/subscribeEventProof.js"
```

can run like:
`yarn validatorRelayer ropsten`

File: scripts/subscribeEventProof.js [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+            lastValidatorsSet.validators = decodedResponse[0].map(validator => validator.toLowerCase());
+            lastValidatorsSet.id = decodedResponse[2].toString();
+            logger.info(`Previous Validator set in bridge is: ${lastValidatorsSet}`);
+        }
+    });
+    return lastValidatorsSet;
+}
+
+async function main (networkName, bridgeContractAddress) {
+    networkName = networkName || 'local';
```
Comment (holygits) [inline scripts/subscribeEventProof.js:117]: pls log the network name, it's useful for debugging

**Review by holygits — APPROVED** [review]

File: scripts/subscribeEventProof.js [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+        user: process.env.supportEmail,
+        pass: process.env.pass
+    }
+});
+
+const mailOptions = {
+    from: process.env.supportEmail,
+    to: process.env.supportEmail,
+    subject: 'Please top up eth balance',
+    text: 'To keep the validator relayer running, topup the eth account'
```
Comment (holygits) [inline scripts/subscribeEventProof.js:20]:
```suggestion
    text: 'To keep the validator relayer running, topup the eth account'
```

can we add the address in this message?

File: package.json (hunk 7 above)
Comment (KarishmaBothara) [reply package.json:18]: doesn't work this way... will try to investigate..

**Pull request merged**